### ISSUE TITLE
Split run list and attributes from nodes

### ISSUE THREAD

**Issue by JanStevens** [higanworks/knife-zero#132]
Hello,

I would like to propose the idea to separate the automatic nodes saving after a chef run from editing the node itself to add attributes or change the run list.

This way you can do a chef run, have a local node file that represents the state of the last run (and allow to easily query with knife) but still have your configuration per node separately in another file.

This would mean using the `--json-attributes` flag (https://docs.chef.io/ctl_chef_client.html) which support run_list / attributes / environment. After the run the state of the node would then be stored in the `nodes` directory.

For example the `json-attributes` file could look like (ex: `servers/dev-server-01.json`):
```js
{
   // Set the run list
  "run_list": ["recipe[base]", "recipe[ci]"], 
  // Specify the chef environment
  "chef_environment": "development",
  // Additional normal attributes
  "ci": {
     "gitlab-runner": {
        "name" : "dev-server-01 bash runner"
      }
  }
}
```

Using the command:

```bash
bundle exec knife zero converge name:dev-server-01 --json-attributes servers/dev-server-01.json
```

After the run the `nodes/dev-server-01.json` file would be updated containing the typical automatic / normal / run list attributes and represents the state of the last run

**Comment by sawanoboly** [comment]
I almost understood the purpose of your proposal.

The `--json-attributes` option already exists and is implemented for the purpose of executing a temporary job with `--override-runlist`.
There is also a `--environment` option, so I think that most of the suggestions can be done with a combination of options.

Getting the environment and run-list from json is a little tricky. I don't want to do much because there are many differences from the core implementation of chef.
How about managing the attachment / detachment of attributes with recipe, role, or data_bag?

**Comment by JanStevens** [comment]
It doesn't really scale well when using recipes / roles or databags since most things are node specific. Having 100 nodes then means 100 recipes which is just a bit annoying.

I don't think that you have to do anything tricky since chef supports passing an json attributes file (including run_list) see: https://docs.chef.io/ctl_chef_client.html

**Comment by sawanoboly** [comment]
Ah, the function was changed. I did not know that. I will look at the code later.

**Comment by sawanoboly** [comment]
I see. https://github.com/chef/chef/blob/master/lib/chef/client.rb#L192

**Comment by sawanoboly** [comment]
@JanStevens 

The node can be saved by overwriting with the exemplified JSON. Can I try this branch?
https://github.com/higanworks/knife-zero/pull/134

**Comment by JanStevens** [comment]
Yes it works exactly how we want it! Thanks a lot for this addition, this greatly improves our workflow with knife-zero

**Comment by sawanoboly** [comment]
released as v2.1.0, Thanks 😃
